File: benchmarks/cpu/mergesort/mergesort.go
package main

import "fmt"

const n = 200
const iterations = 50

func nextRand(seed int64) int64 {
	s := seed*1103515245 + 12345
	if s < 0 {
		s = -s
	}
	s = s % 1000000007
	return s
}

func merge(left, right []int64) []int64 {
	result := make([]int64, 0, len(left)+len(right))
	i, j := 0, 0
	for i < len(left) && j < len(right) {
		if left[i] <= right[j] {
			result = append(result, left[i])
			i++
		} else {
			result = append(result, right[j])
			j++
		}
	}
	result = append(result, left[i:]...)
	result = append(result, right[j:]...)
	return result
}

func mergesort(arr []int64) []int64 {
	if len(arr) <= 1 {
		sorted := make([]int64, len(arr))
		copy(sorted, arr)
		return sorted
	}
	mid := len(arr) / 2
	left := mergesort(arr[:mid])
	right := mergesort(arr[mid:])
	return merge(left, right)
}

func main() {
	original := make([]int64, n)
	seed := int64(42)
	for i := 0; i < n; i++ {
		seed = nextRand(seed)
		original[i] = seed % 10000
	}

	var firstElement int64
	for iter := 0; iter < iterations; iter++ {
		sorted := mergesort(original)
		firstElement = sorted[0]
	}

	fmt.Println(firstElement)
}

File: benchmarks/cpu/mergesort/mergesort.py
N = 200
ITERATIONS = 50


def next_rand(seed):
    s = seed * 1103515245 + 12345
    if s < 0:
        s = -s
    s = s % 1000000007
    return s


def merge(left, right):
    result = []
    i = 0
    j = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    while i < len(left):
        result.append(left[i])
        i += 1
    while j < len(right):
        result.append(right[j])
        j += 1
    return result


def mergesort(arr):
    if len(arr) <= 1:
        return arr[:]
    mid = len(arr) // 2
    left = mergesort(arr[:mid])
    right = mergesort(arr[mid:])
    return merge(left, right)


def main():
    original = []
    seed = 42
    for _ in range(N):
        seed = next_rand(seed)
        original.append(seed % 10000)

    first_element = 0
    for _ in range(ITERATIONS):
        sorted_arr = mergesort(original)
        first_element = sorted_arr[0]

    print(first_element)


if __name__ == "__main__":
    main()
